Add information about config files of async api (#7654)
magento/devdocs: src/guides/v2.3/extension-dev-guide/message-queues/async-message-queue-config-files.md

---
group: php-developer-guide
title: Asynchronous Message Queue configuration files
contributor_name: comwrap GmbH
contributor_link: https://www.comwrap.com
functional_areas:
  - Services
---

For user Magento message queue, you have to utilise four configuration files in your module:

*  communication.xml
*  queue_consumer.xml
*  queue_topology.xml
*  queue_publisher.xml

More about them can be found in [Configure message queues]({{ page.baseurl }}/guides/v2.4/extension-dev-guide/message-queues/config-mq.html).

> Asynchronous and Bulk APIs are build on top of usuall REST API and using Magento Message Queue Framework for processing messages. For make customer development as less as possible and make integration as easy as possible, Asynchronous API pre-generating following configuration files: 

Reviewer: ```suggestion
Asynchronous and Bulk APIs are built on top of the usual REST API and using Magento Message Queue Framework for processing messages. For making development effort as less as possible and making integration as easy as possible, Asynchronous API pre-generating following configuration files: 
```

*  communication.xml
*  queue_publisher.xml

> And provides `queue_topology.xml` and `queue_consumer.xml` within Magento/WebapiAsync module.

Reviewer: ```suggestion
and provides `queue_topology.xml` and `queue_consumer.xml` within `Magento/WebapiAsync` module.
```

### communication.xml

Information about auto-generation of `communication.xml` can be found in [Topics in Asynchronous API]({{ page.baseurl }}/guides/v2.4/extension-dev-guide/message-queues/async-topics.md)

### queue_publisher.xml

`queue_publisher.xml` is generated by `\Magento\WebapiAsync\Code\Generator\Config\RemoteServiceReader\Publisher` class, which implements `\Magento\Framework\Config\ReaderInterface` and injected into `\Magento\Framework\MessageQueue\Publisher\Config\CompositeReader` constructor argument of main `di.xml` file.

```xml
<type name="Magento\Framework\MessageQueue\Publisher\Config\CompositeReader">
    <arguments>
        <argument name="readers" xsi:type="array">
            <item name="asyncServiceReader" xsi:type="object" sortOrder="0">Magento\WebapiAsync\Code\Generator\Config\RemoteServiceReader\Publisher</item>
            ...
        </argument>
    </arguments>
</type>
```

Sort order is set to 0 and this allows developers to change some aspects of generated configuration in the config readers such as `queue_publisher.xml`, `env.php`.

> `\Magento\WebapiAsync\Code\Generator\Config\RemoteServiceReader\Publisher::read()` calls `\Magento\AsynchronousOperations\Model\ConfigInterface::getServices()` to get an array of all REST API routes which could be executed asynchronously. Then it defines connection name equals to `amqp` and echange equals to `magento` for each generated topic name.

Reviewer: ```suggestion
`\Magento\WebapiAsync\Code\Generator\Config\RemoteServiceReader\Publisher::read()` calls `\Magento\AsynchronousOperations\Model\ConfigInterface::getServices()` to get an array of all REST API routes which could be executed asynchronously. Then it defines connection name equals to `amqp` and exchange equals to `magento` for each generated topic name.
```
